server/src/test/java/org/elasticsearch/indices/recovery/plan/SnapshotsRecoveryPlannerServiceTests.java

```diff
@@ -180,23 +184,37 @@ public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<Optiona
             assertAllSourceFilesAreAvailableInSource(shardRecoveryPlan, sourceMetadata);
             assertAllIdenticalFilesAreAvailableInTarget(shardRecoveryPlan, targetSourceMetadata);
             assertUsesExpectedSnapshot(shardRecoveryPlan, shardSnapshotData);
+            assertThat(shardRecoveryPlan.canRecoverSnapshotFilesFromSourceNode(), is(equalTo(true)));
 
             assertThat(shardRecoveryPlan.getStartingSeqNo(), equalTo(startingSeqNo));
             assertThat(shardRecoveryPlan.getTranslogOps(), equalTo(translogOps));
         });
     }
 
-    public void testLogicallyEquivalentSnapshotIsSkippedIfUnderlyingFilesAreDifferent() throws Exception {
+    public void testLogicallyEquivalentSnapshotIsUsedEvenIfFilesAreDifferent() throws Exception {
         createStore(store -> {
-            Store.MetadataSnapshot targetSourceMetadata = generateRandomTargetState(store);
+            boolean shareFilesWithSource = randomBoolean();
+            Store.MetadataSnapshot targetSourceMetadata = generateRandomTargetState(store, shareFilesWithSource);
 
             writeRandomDocs(store, randomIntBetween(10, 100));
             Store.MetadataSnapshot sourceMetadata = store.getMetadata(null);
 
+
+            boolean compatibleVersion = randomBoolean();
+            final Version snapshotVersion;
+            final org.apache.lucene.util.Version luceneVersion;
+            if (compatibleVersion) {
+                snapshotVersion = randomBoolean() ? null : randomCompatibleVersion(random(), Version.CURRENT);
+                luceneVersion = randomCompatibleVersion(random(), Version.CURRENT).luceneVersion;
```

Comment:
In the case where `snapshotVersion == null` we need to use a `luceneVersion` that is before the lucene version of `SEQ_NO_SNAPSHOT_RECOVERIES_SUPPORTED_VERSION`